cirq/sim/state.py

```diff
@@ -121,3 +123,139 @@ def validate_normalized_state(state: np.ndarray, num_qubits: int,
+
+    Args:
+        state: The state to be measured. This state is assumed to be normalized.
+            The state must be of size 2 ** integer.  The state can be of shape
+            (2 ** integer) or (2, 2, ..., 2).
+        indices: Which qubits are measured. The state is assumed to be supplied
+            in big endian order. That is the xth index of v, when expressed as
+            a bitstring, has the largest values that the 0th index.
+
+    Returns:
+        A tuple of a list and an numpy array. The list is an array of booleans
+        corresponding to the measurement values (ordered by the indices). The
+        numpy array is the post measurement state. This state has the same
+        shape as the input state.
+
+    Raises:
+        ValueError if the size of state is not a power of 2.
+        IndexError if the indices are out of range for the number of qubits
+            corresponding to the state.
+    """
+    num_qubits = _validate_num_qubits(state)
+    _validate_indices(num_qubits, indices)
+
+    if len(indices) == 0:
+        return ([], np.copy(state))
+
+    # Calculate the measurement probabilities and then make the measurement.
+    probs = _probs(state, indices, num_qubits)
+    result = np.random.choice(2 ** len(indices), p=probs)
+    measurement_bits = [(1 & (result >> i)) for i in range(len(indices))]
+
+    # Calculate the slice for the measurement result.
+    b = sum(2**i*b for i, b in enumerate(measurement_bits))
+    result_slice = linalg.slice_for_qubits_equal_to(indices, b)
+
+    # Create a mask which is False for only the slice.
+    mask = np.ones([2] * num_qubits, dtype=bool)
+    mask[result_slice] = False
+
+    # Potentially reshape to tensor, and then set masked values to 0.
+    tensor = np.reshape(np.copy(state), [2] * num_qubits)
+    tensor[mask] = 0
+
+    # Restore original shape (if necessary) and renormalize.
+    return_state = np.reshape(tensor, state.shape) / np.sqrt(probs[b])
+
+    return (measurement_bits, return_state)
+
+
+def _probs(state: np.ndarray, indices: List[int],
+        num_qubits: int):
+    """Returns the probabilities for a measurement on the given indices."""
+    # Tensor of squared amplitudes, shaped a rank [2, 2, .., 2] tensor.
+    prob_tensor = np.abs(np.reshape(state, [2] * num_qubits)) ** 2
+
+    # Calculate the probabilities for measuring the particular results.
+    probs = [np.sum(prob_tensor[linalg.slice_for_qubits_equal_to(indices, b)])
+             for b in range(2 ** len(indices))]
+    # To deal with rounding issues, ensure that the probabilities sum to 1.
+    probs /= sum(probs) # type: ignore
+    return probs
+
+
+def _validate_num_qubits(state: np.ndarray) -> int:
+    """Validates that state's size is a power of 2, returning number of qubits.
+    """
+    size = state.size
+    if (size & (size - 1)):
+        raise ValueError(
+            'State is not of size a power of two instead was {}'.format(size))
```

Comment:
`"state.size ({}) isn't a power of two"`

Reply:
done

Reply:
Looks not done.